### PR TITLE
linkTypeTopology

### PR DESCRIPTION
**Pull request by plutoshe**
Initial interface for linkTopology.
Substitute the topology with a set of linkTopology.
For instance, we describe a topology as a set of linkType, and corresponding sets of pointed linkType before. After changing to linkType, for every linkType, we describe as a topology.
It's more flexible to fit the latter structure.

### CODE REVIEW

**Comment by plutoshe** [comment]
Let's discuss details.

File: framework/bootstrap.go [hunk 1]
```diff
@@ -28,7 +28,9 @@ func (f *framework) SetTaskBuilder(taskBuilder taskgraph.TaskBuilder) {
 	f.taskBuilder = taskBuilder
 }
 
-func (f *framework) SetTopology(topology taskgraph.Topology) { f.topology = topology }
+func (f *framework) AddLinkage(topology taskgraph.Topology) {
```
Comment (fengjingchao) [inline framework/bootstrap.go]: Why you change the name? Can you add a comment to the function?

File: framework/bootstrap.go [hunk 2]
```diff
@@ -63,7 +65,9 @@ func (f *framework) Start() {
 	// Both should be initialized at this point.
 	// Get the task implementation and topology for this node (indentified by taskID)
 	f.task = f.taskBuilder.GetTask(f.taskID)
-	f.topology.SetTaskID(f.taskID)
+	for key, _ := range f.topology {
```
Comment (fengjingchao) [inline framework/bootstrap.go]: Can you explain the change and also change the comment?

File: framework/framework.go [hunk 3]
```diff
@@ -22,7 +22,7 @@ type framework struct {
 
 	// user defined interfaces
 	taskBuilder taskgraph.TaskBuilder
-	topology    taskgraph.Topology
+	topology    map[string]taskgraph.Topology
```
Comment (fengjingchao) [inline framework/framework.go]: Can you explain by adding comments?

File: framework_interface.go [hunk 4]
```diff
@@ -14,7 +14,7 @@ type Bootstrap interface {
 	SetTaskBuilder(taskBuilder TaskBuilder)
 
 	// This allow the application to specify how tasks are connection at each epoch
-	SetTopology(topology Topology)
+	AddLinkage(linkType string, topology Topology)
```
Comment (fengjingchao) [inline framework_interface.go]: Explain and change comment.

**Comment by fengjingchao** [comment]
We need to revert this since Xiaoyun took it as #121 .

**Comment by fengjingchao** [comment]
Pluto, can you submit another pull request and also compress the commits to the last one? That's the only commit passed testing.

**Pull request merged**